File: backend/app/schemas/blockchain_explorer.py
from uuid import UUID

from pydantic import BaseModel, Field


class BlockchainExplorerUser(BaseModel):
    user_id: UUID
    badge_number: str | None = None
    username: str | None = None
    email: str | None = None
    full_name: str | None = None
    rank: str | None = None


class BlockchainOverviewResponse(BaseModel):
    network: str
    consensus: str
    enabled: bool
    connected: bool
    chain_id: int | None
    latest_block: int | None
    contract_address: str | None
    contract_deployed: bool
    deployment_block: int


class BlockchainTransactionSummary(BaseModel):
    tx_hash: str
    from_address: str | None = None
    to_address: str | None = None
    transaction_index: int
    is_registry_transaction: bool


class BlockchainBlockResponse(BaseModel):
    block_number: int
    block_hash: str
    timestamp: int
    parent_hash: str
    transaction_count: int
    transactions: list[BlockchainTransactionSummary] = Field(default_factory=list)


class BlockchainRegistryEvent(BaseModel):
    event_type: str
    evidence_ref: str
    evidence_hash: str | None = None
    uploader_ref: str | None = None
    officer_ref: str | None = None
    access_session_ref: str | None = None
    action: str | None = None
    occurred_at: int | None = None
    recorded_at: int
    writer: str
    tx_hash: str
    block_number: int
    transaction_index: int
    log_index: int


class BlockchainTransactionResponse(BaseModel):
    tx_hash: str
    status: str
    block_number: int
    from_address: str | None = None
    to_address: str | None = None
    transaction_index: int
    gas_used: int
    contract_address: str | None = None
    is_registry_transaction: bool
    registry_events: list[BlockchainRegistryEvent] = Field(default_factory=list)


class BlockchainEvidenceRegistration(BaseModel):
    evidence_hash: str
    uploader_ref: str
    recorded_at: int
    writer: str
    tx_hash: str | None = None
    block_number: int | None = None
    transaction_index: int | None = None
    log_index: int | None = None


class BlockchainAccessEvent(BaseModel):
    evidence_ref: str
    officer_ref: str
    access_session_ref: str
    action: str
    occurred_at: int
    recorded_at: int
    writer: str
    tx_hash: str
    block_number: int
    transaction_index: int | None = None
    log_index: int | None = None
    actor: BlockchainExplorerUser | None = None


class BlockchainEvidenceResponse(BaseModel):
    evidence_id: UUID | None = None
    evidence_number: str | None = None
    evidence_ref: str
    registration: BlockchainEvidenceRegistration
    access_history: list[BlockchainAccessEvent] = Field(default_factory=list)
    scan_from_block: int
    scan_to_block: int


class BlockchainAccessSessionResponse(BaseModel):
    access_session_ref: str
    evidence_ref: str
    officer_ref: str
    action: str
    occurred_at: int
    recorded_at: int
    writer: str
    tx_hash: str | None = None
    block_number: int | None = None
    transaction_index: int | None = None
    log_index: int | None = None
    evidence_id: UUID | None = None
    evidence_number: str | None = None
    actor: BlockchainExplorerUser | None = None
    database_access_log_found: bool = False
    database_access_log_id: UUID | None = None

File: frontend/src/interfaces/blockchainExplorer.ts
export type BlockchainSearchType =
  | "block"
  | "transaction"
  | "evidence"
  | "evidence-ref"
  | "access-session";

export interface BlockchainOverview {
  network: string;
  consensus: string;
  enabled: boolean;
  connected: boolean;
  chain_id: number | null;
  latest_block: number | null;
  contract_address: string | null;
  contract_deployed: boolean;
  deployment_block: number;
}

export interface BlockchainTransactionSummary {
  tx_hash: string;
  from_address: string | null;
  to_address: string | null;
  transaction_index: number | null;
  is_registry_transaction: boolean;
}

export interface BlockchainBlockResult {
  block_number: number;
  block_hash: string;
  timestamp: number;
  parent_hash: string;
  transaction_count: number;
  transactions: BlockchainTransactionSummary[];
}

export interface BlockchainRegistryEvent {
  event_type: string;
  evidence_ref: string;
  evidence_hash: string | null;
  uploader_ref: string | null;
  officer_ref: string | null;
  access_session_ref: string | null;
  action: string | null;
  occurred_at: number | null;
  recorded_at: number;
  writer: string;
  tx_hash: string;
  block_number: number;
  transaction_index: number | null;
  log_index: number | null;
  evidence_id: string | null;
  evidence_number: string | null;
  actor: BlockchainExplorerUser | null;
  uploader: BlockchainExplorerUser | null;
  related_evidence: BlockchainRelatedEvidence | null;
}

export interface BlockchainTransactionResult {
  tx_hash: string;
  status: string;
  block_number: number;
  from_address: string | null;
  to_address: string | null;
  transaction_index: number | null;
  gas_used: number;
  contract_address: string | null;
  is_registry_transaction: boolean;
  registry_events: BlockchainRegistryEvent[];
}

export interface BlockchainExplorerUser {
  user_id: string;
  badge_number: string | null;
  username: string | null;
  email: string | null;
  full_name: string | null;
  rank: string | null;
}

export interface BlockchainRelatedEvidence {
  evidence_hash: string;
  uploader_ref: string;
  recorded_at: number;
  writer: string;
  uploader: BlockchainExplorerUser | null;
}

export interface BlockchainAccessEvent {
  evidence_ref: string;
  officer_ref: string;
  access_session_ref: string;
  action: string;
  occurred_at: number;
  recorded_at: number;
  writer: string;
  tx_hash: string;
  block_number: number;
  transaction_index: number | null;
  log_index: number | null;
  actor: BlockchainExplorerUser | null;
}

export interface BlockchainEvidenceResult {
  evidence_id: string | null;
  evidence_number: string | null;
  evidence_ref: string;
  registration: {
    evidence_hash: string;
    uploader_ref: string;
    recorded_at: number;
    writer: string;
    tx_hash: string | null;
    block_number: number | null;
    transaction_index: number | null;
    log_index: number | null;
    uploader: BlockchainExplorerUser | null;
  };
  access_history: BlockchainAccessEvent[];
  scan_from_block: number;
  scan_to_block: number;
}

export interface BlockchainAccessSessionResult {
  access_session_ref: string;
  evidence_ref: string;
  officer_ref: string;
  action: string;
  occurred_at: number;
  recorded_at: number;
  writer: string;
  tx_hash: string | null;
  block_number: number | null;
  transaction_index: number | null;
  log_index: number | null;
  evidence_id: string | null;
  evidence_number: string | null;
  actor: BlockchainExplorerUser | null;
  related_evidence: BlockchainRelatedEvidence | null;
  database_access_log_found: boolean;
  database_access_log_id: string | null;
}

export type BlockchainSearchResult =
  | { type: "block"; data: BlockchainBlockResult }
  | { type: "transaction"; data: BlockchainTransactionResult }
  | { type: "evidence" | "evidence-ref"; data: BlockchainEvidenceResult }
  | { type: "access-session"; data: BlockchainAccessSessionResult };
